### DISCUSSION TITLE
Importing type variants

### DISCUSSION THREAD

**Discussion by Dan-C-Reed** [gleam-lang/gleam#3232] [Questions & support]
I have a file called `types.gleam` that contains the following:

```gleam
pub type DataSource {
  Url
  File
}
```
and I'd like to use it in another file, so I import the `DataSource` type like so:

```gleam
import gleam/string
import types.{type DataSource}

pub fn get_source_type(path: String) -> DataSource {
  case string.starts_with(path, "http") {
    True -> Url
    False -> File
  }
}
```

This throws an error that `Url` is not in scope despite importing `DataSource`. I think I must also need to import the `Url` variant of `DataSource`, however, I'm not sure how. When the function and the type live in the same file everything is peachy. 

Can someone advise me how to import a variant of a type?

**Comment by Dan-C-Reed** [comment]
Of course, immediately after I post this I figure it out. :man_facepalming:  I import `Url` like a function and not a type because it's considered a constructor. So,

```gleam
import types.{type DataSource, Url, File}
```

**Comment by barcek** [comment]
Adding a thought on this subject, re something I've noticed early on using Gleam. Importing a larger number of variants in this way can get unwieldy, especially having to update imports if the custom type is extended or reduced at a later point in development.

Using @Dan-C-Reed's example above, it would be useful if all variants didn't need to be imported, but could be accessed via the type in place, e.g.:

```gleam
import types.{type DataSource}
...
  True -> DataSource.Url
...
```

This would help clarify the relationship between variants locally too.

Using the qualified import syntax, i.e. accessing via the module name, is possible and reasonable if there's only one custom type per module, but with multiple types each with multiple variants also potentially confusing.

**Reply by barcek** [reply]
Didn't spot this was closed, and wouldn't show up in the default listing. Posted with some changes as discussion [#3512](https://github.com/gleam-lang/gleam/discussions/3512).
